README.md
# Vampires
week 6 Lab-Game


history
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>History</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" media="screen" href="main.css" />
    <script src="main.js"></script>
</head>
<body>
    
</body>
</html>

js/src/main.js
//MAIN
function update() {
  frames++;
  // if(frames==1){
  //   vamp.sprite = new Sprite('images/Demitri.png',[0,8945],[205,160 ],12,[0,1,2,3,4,5,6,7,8,9,10],null,false,ctx, vamp.xpoint, vamp.ypoint, vamp.width, vamp.height);
  
  // }
  //ctx.clearRect(0, 0, canvas.width, canvas.height);
  board.drawFloor();
  board.draw();
  board.drawScore();
  vamp.draw();
  vamp.verticalStrike();
  vamp.horizontalStrike();
 
  switch (moves){
    case 0:
      vamp.idle();
      break;
    case 1:
      vamp.punch();
      vamp.soundpunch.play();
      break;
    case 2:
      vamp.kick();
      vamp.soundkick.play();
      break;
    case 3:
      vamp.power();
      fires.sound.play();
      break;
    case 4:
      vamp.moveRight();
      break;
    case 5:
      vamp.moveLeft();
      break;
    case 6:
      vamp.moveDown();
      break;
    case 7:
      vamp.moveUp();
      break;
    
  }
  drawEnemies();
  drawFires();
  generateEnemies();
  checkCollition();
  //hitDetection();
  //if(frames%160){
  //vamp.sprite.update(vamp,frames);
  //vamp.sprite.render(ctx);
  //}
}

function startGame() {
  //vamp.idle();
  //moves = 0;
  if (intervalo > 0) return;

  intervalo = setInterval(function() {
    update();
  }, 1000 / 60);
  // intervalo =  requestAnimationFrame(function () {
  //  update();
  // },1000/60)
  board.sound.play();
  board.sound2.play();
  
}

// var lastTime;
// function startGame(){
//   var now = Date.now();
//   var dt = (now - lastTime) / 1000.0;
//   update(dt);
//   //render();
//   lastTime = now;
//   requestAnimFrame(startGame);
//   board.sound.play();
//   board.sound2.play();
// }

function stopGame() {
  clearInterval(intervalo);
  intervalo = 0;
 board.sound.pause();
 board.sound3.play();
}

//funcion de popup con html index

//amplio el vcanvas con boton comunica con html
document.getElementById("btnAmpliar").onclick = function() {
  ampliar();
};

function ampliar() {
  contenedor.style.width = "100%";
  contenedor.style.height = "100vh";
  contenedor.style.margin = "0";

  ampliarCanvas.style.width = "100%";
  ampliarCanvas.style.height = "100vh";
  ampliarCanvas.style.backgroundSize = "cover";
  ampliarCanvas.style.backgroundRepeat = "no-repeat";

  btnAmpliar.innerHTML = "reducir";
  btnAmpliar.style.position = "fixed";
  btnAmpliar.style.top = "10px";
  btnAmpliar.style.left = "10px";
  btnAmpliar.style.zIndex = "1";
  btnStart.style.display = "none";
  btnReducir.style.display = "none";
  btnAmpliar.setAttribute("onClick", "reducir()");
}

function reducir() {
  contenedor.style.width = "1400px";
  contenedor.style.height = "700px";
  contenedor.style.margin = "5vh auto";

  ampliarCanvas.style.width = "1400px";
  ampliarCanvas.style.height = "700px";

  btnAmpliar.innerHTML = "pantalla full";
  btnAmpliar.style.position = "relative";
  btnAmpliar.style.top = "0";
  btnAmpliar.style.left = "0";
  btnAmpliar.style.zIndex = "0";
  btnStart.style.display = "inline";
  btnReducir.style.display = "inline";
  btnAmpliar.setAttribute("onClick", "ampliar()");
}


js/src/gamepad.js


// function gamepadHandler(event, connecting) {
//   var gamepad = event.gamepad;
//   // Note:
//   // gamepad === navigator.getGamepads()[gamepad.index]

//   if (connecting) {
//     gamepads[gamepad.index] = gamepad;
//   } else {
//     delete gamepads[gamepad.index];
//   }
// }

// var hasGP = false;
// var repGP;
// function canGame() {
//   return "getGamepads" in navigator;
// }
// var input = {
//   left: false,
//   right: false
// };
// $(window).keydown(function(e) {
//   switch (e.keyCode) {
//        case 37: input.left = true; break;                            
//        case 39: input.right = true; break;                            
//   } 
// });
// $(window).keyup(function(e) {
//   switch (e.keyCode) {
//        case 37: input.left = false; break;                            
//        case 39: input.right = false; break;                            
//   } 
// });

// this.move = function() {
//   if(input.left) {
//     this.x -= this.speed;
//     if(this.x < 0) this.x=0;
//   }
//   if(input.right) {
//     this.x += this.speed;
//     if((this.x+this.width) > canvas.width) this.x=canvas.width-this.width;
//   }
// }


// //gamepad support
// if(canGame()) {
//   var prompt = "To begin using your gamepad, connect it and press any button!";
//   $("#gamepadPrompt").text(prompt);
//   $(window).on("gamepadconnected", function() {
//     hasGP = true;
//     $("#gamepadPrompt").html("Gamepad connected!");
//     repGP = window.setInterval(checkGamepad,100);
//   });
//   $(window).on("gamepaddisconnected", function() {
//     $("#gamepadPrompt").text(prompt);
//     window.clearInterval(repGP);
//   });
//   //setup an interval for Chrome
//   var checkGP = window.setInterval(function() {
//     if(navigator.getGamepads()[0]) {
//       if(!hasGP) $(window).trigger("gamepadconnected");
//       window.clearInterval(checkGP);
//     }
//   }, 500);
  
//   function checkGamepad() {
//     var gp = navigator.getGamepads()[0];
//     var axeLF = gp.axes[0];
//     if(axeLF < -0.5) {
//       input.left = true;
//       input.right = false;
//     } else if(axeLF > 0.5) {
//       input.left = false;
//       input.right = true;
//     } else {
//       input.left = false;
//       input.right = false;
//     }
//   }
// }		

js/src/sprites.js
function Sprite(
  url,
  pos,
  size,
  speed,
  frames,
  dir,
  once,
  ctx,
  x,
  y,
  width,
  height
) {
  this.pos = pos;
  this.size = size;
  this.speed = typeof speed === "number" ? speed : 0;
  this.frames = frames;
  this._index = 0;
  this.url = url;
  this.dir = dir || "horizontal";
  this.once = once;
  this.width = width;
  this.height = height;
  this.x = x;
  this.y = y;
  this.img = new Image();
  this.img.src = url;

  this.draw = function(ctx, x, y, width, height) {
    ctx.drawImage(
      this.img,
      x,
      y,
      this.size[0],
      this.size[1],
      this.x,
      this.y,
      this.width,
      this.height
    );
  };
}

Sprite.prototype = {
  // update: function(dt) {
  //     this._index += this.speed*dt;
  // },

  update: function(vamp, fra) {
    if (fra % 5 == 0) {
      this._index++;
      this.y = vamp.ypoint;
      this.x = vamp.xpoint;
      this.width = vamp.Dwidth;
      this.height = vamp.Dheight;
    }
  },
  render: function(ctx) {
    var frame;

    if (this.speed > 0) {
      var max = this.frames.length;
      var idx = Math.floor(this._index);
      frame = this.frames[idx % max];

      if (this.once && idx >= max) {
        moves = 0;
        this._index = 0;
        this.done = true;
        return;
      }
    } else {
      //moves = 0;
      frame = 0;
    }

    var x = this.pos[0];
    var y = this.pos[1];

    //if(this.dir == 'vertical') {
    //  y += frame * this.size[1];
    ///  }
    //else {
    x += frame * this.size[0];
    //  }

    this.draw(ctx, x, y, this.width, this.height);
    // ctx.drawImage(this.image,
    //               x, y,
    //               this.size[0], this.size[1],
    //               0, 0,
    //               this.size[0], this.size[1]);
  }
};

window.Sprite = Sprite;


js/src/listeners.js
document.getElementById("start-button").onclick = function() {
  startGame();
};

addEventListener("keydown", function(e) {
  switch (e.keyCode) {
    case 87:
      moves = 7;
      //vamp.moveUp();
      //console.log("up", vamp);
      break;
    case 83:
      moves = 6;
      //vamp.moveDown();
      //console.log("down", vamp);
      break;
    case 65:
      moves = 5;
      //vamp.moveLeft();
      //console.log("left", vamp);
      break;
    case 68:
      moves = 4;
      //vamp.moveRight();
      //console.log("right", vamp);
      break;
    case 38:
      moves = 1;
     //vamp.punch();
     //console.log("punch", moves);
      break;
    case 39:
      moves = 2;
      //vamp.kick();
      //moves = false;
     // console.log("kick", moves);
      break;
    case 40:
      moves = 3;
      generateFires();
      
      //vamp.power();
      //console.log("pow", vamp);
      break;
  }
});




js/src/auxiliares.js
//generate
function generateEnemies() {
  if (!(frames % 30 === 0)) return;
  var randomPosition = Math.floor(Math.random() * 399);
  var randomEnemie = Math.floor(Math.random() * 4);
  if (randomPosition > 170 && randomPosition < 350) {
    //rango de area de juego solo ejecute enemigos entro del area jugable
    var enemie = new Enemies(randomPosition, randomEnemie);
    enemiesArray.push(enemie);
  }

  // console.log("de: "+randomPosition);
  // console.log("le toca salir a : "+randomEnemie);
  // console.log("we are in generate enemies function");
}

function generateFires() {
  var fires = new Fire(vamp);
  fireArray.push(fires);
  console.log(fireArray);
}

function drawFires() {
  fireArray.forEach(function(fires) {
    fires.draw();
  });
}

function drawEnemies() {
  enemiesArray.forEach(function(enemie) {
    enemie.draw();
  });
}

function killEnemies(enemie) {
  enemiesArray.splice(enemie);
}

function checkCollition() {
  //console.log("Moves:", moves);
  enemiesArray.forEach(function(enemie, indice) {
    if (vamp.isTouching(enemie) && moves == 0) {
      vamp.health -= 25;
      enemiesArray.splice(indice, 1);
    } else if (vamp.isTouching(enemie) && moves > 0) {
      enemiesArray.splice(indice, 1);
      //killEnemies();
    } else if (vamp.health == 0) {
      gameOver();
    }
  });

}

function hitDetection(){
  enemiesArray.forEach(function(enemie,indice){
    fireArray.forEach(function(fires){
      if(enemie.isTouching(fires)){
        console.log("dead to all");
        enemiesArray.splice(indice,1);
        vamp.health +=25;
        vamp.soundrisa.play();
      }
    })
  })
}

function gameOver() {
  stopGame();
  ctx.font = " 160px Unquiet Spirits";
  ctx.strokeStyle = "red";
  ctx.lineWidth = 8;
  ctx.strokeText("Game Over", 500, 300);

  //ctx.font = '50px Avenir';
}


js/src/declarations.js
//declaraciones

//canvas variables.
var canvas = document.getElementById("canvas");
var ctx = canvas.getContext("2d");
var intervalo;
var frames = 0;
var board = new Board();
//heroes

var vamp = new Vamp();
var enemie = new Enemies(fires);
var fires = new Fire(vamp);
var moves = 0;

var enemiesArray = [];
var fireArray = [];
//animation
var requestAnimFrame = (function() {
  return (
    window.requestAnimationFrame ||
    window.webkitRequestAnimationFrame ||
    window.mozRequestAnimationFrame ||
    window.oRequestAnimationFrame ||
    window.msRequestAnimationFrame ||
    function(callback) {
      window.setTimeout(callback, 1000 / 60);
    }
  );
})();
//sprites

//seleccion de DOM
var contenedor = document.querySelector("#contenedor");
var ampliarCanvas = document.querySelector("#canvas");
var btnStart = document.querySelector("#start-button");
var btnReducir = document.querySelector("#btnReturn");
var output = document.querySelector("#output");

//GAMEPAD request animation frame
var rAF =
  window.mozRequestAnimationFrame ||
  window.webkitRequestAnimationFrame ||
  window.requestAnimationFrame;
var rAFStop =
  window.mozCancelRequestAnimationFrame ||
  window.webkitCancelRequestAnimationFrame ||
  window.cancelRequestAnimationFrame;
//gamepads
var gamepad = {};
var gamepads = navigator.getGamepads();
var controller = {};
var buttonsPressed = [];


js/src/classes.js
//classes

//demitri
function Vamp() {
  this.health = 200;
  this.xpoint = 100;
  this.ypoint = 300;
  this.speed = 3;
  this.Dwidth = 500; //250; //833/8 = 110
  this.Dheight = 500; //500; //181
  //sonidos
  this.soundpunch = new Audio();
  this.soundpunch.src = 'images/sonidos/punch.wav';
  this.soundkick = new Audio();
  this.soundkick.src = 'images/sonidos/patada.wav';
  this.soundrisa = new Audio();
  this.soundrisa.src = 'images/sonidos/risa.wav';
  //movimiento
  this.ydir = 5;
  this.xdir = 5;
  // this.img = new Image();
  // this.img.src = 'images/Demitri_mod.png'
  //sprite

  this.idleSprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 281],
    [180,140],
    7,
    [1,2,3,4,5],
    null,
    false,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.punchingSprite = new Sprite(
    "images/DemitripunchLX.png",
    [0, 0],
    [226.77, 143],
    7,
    [0, 1, 2, 3, 4, 5, 6, 7, 8],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.walksprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 844],
    [180, 140],
    9,
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.walkLeftSprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 133],
    [180, 140],
    9,
    [0, 1, 2, 3, 4, 5, 6, 7],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.kickSprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 427],
    [180, 140],
    8,
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 11, 12],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.powerSprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 565],
    [180, 140],
    20,
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.width,
    this.height
  );

  this.onload = function () {
    this.draw;
    //ctx.fillRect(this.xpoint, this.ypoint, this.Dwidth, this.Dheight);
  }.bind(this);

  this.draw = function () {
    ctx.font = " 90px Unquiet Spirits";
    ctx.strokeText("Vampire", 150, 90);
    ctx.strokeStyle = "purple";
    ctx.fillStyle = "#FF0000";
    ctx.fillRect(400, 30, (this.health / 200) * 200, 25);


  }; //draw

  this.verticalStrike = function () {
    if (this.ypoint <= 170) {
      this.ypoint = 200 + this.ydir;
    }
    if (this.ypoint >= 500) {
      this.ypoint = 470 + this.ydir;
    }
  };

  this.horizontalStrike = function () {
    if (this.xpoint <= 0) {
      this.xpoint = 20 + this.xdir;
    }
    if (this.xpoint >= canvas.width) {
      this.xpoint = 1200 + this.xdir;
    }
  };
  this.idle = function () {
    //ctx.drawImage(this.img, this.xpoint, this.ypoint, 250,200);
    //this.image = this.img;

    this.sprite = this.idleSprite;
    this.sprite.update(this, frames);
   this.sprite.render(ctx);
    
  }
  this.punch = function () {
    this.sprite = this.punchingSprite;
    this.sprite.update(this, frames);
    this.sprite.render(ctx);
    this.soundpunch;
    //this.sprite.
  };

  this.moveRight = function () {
    board.moveForwad();
    this.sprite = this.walksprite;
    this.xpoint += this.xdir;// + this.speed; //duda de seguimiento
    this.sprite.update(vamp, frames);
    this.sprite.render(ctx);
  };

  this.moveLeft = function () {
    this.sprite = this.walkLeftSprite;
    this.sprite.update(vamp, frames);
    this.sprite.render(ctx);
    this.xpoint -= this.xdir;// + this.speed;
    
    
  };

  this.moveUp = function () {
    this.sprite = this.walksprite;
    this.sprite.update(this, frames);
    this.sprite.render(ctx);
    this.ypoint -= this.ydir;// + this.speed;
  };
  this.moveDown = function () {
    this.sprite = this.walksprite;
    this.sprite.update(this, frames);
    this.sprite.render(ctx);
    this.ypoint += this.ydir;// + this.speed;
  };


  //limite de movimiento hasta y175;

  this.kick = function () {
    this.sprite = this.kickSprite;
    this.sprite.update(vamp,frames);
    this.sprite.render(ctx);
    

  };

  this.power = function () {
    this.sprite = this.powerSprite;
    this.sprite.update(vamp,frames);
    this.sprite.render(ctx);
    
    
  };

  this.isTouching = function (enemie) {
   
    return (this.xpoint - 90 < enemie.x + enemie.width) && //si heroe es menor que enemigo + width
      (this.xpoint - 90 + this.Dwidth - 90 > enemie.x) && //si herore + su ancho es mayor que el alto del enemigo
      (this.ypoint + 100 < enemie.y + enemie.height) && //si el alto del heroe es menor que el alto del enemifo mas su alto
      (this.ypoint + 100 + this.Dheight + 150 > enemie.y);
    // (this.x < pipe.x + pipe.width) &&
    // (this.x + this.width > pipe.x) &&
    // (this.y < pipe.y + pipe.height) &&
    // (this.y + this.height > pipe.y);
  };
} //final Dimitri

function Fire(vamp) {
  this.speed = 200;
  this.xpoint =vamp.xpoint + vamp.Dwidth/1.3;
  this.ypoint = (vamp.ypoint + vamp.Dheight)/2;
  this.Dwidth = 350;
  this.Dheight = 350;
  this.sound = new Audio();
  this.sound.src = 'images/sonidos/pow.wav';
  this.sound2 = new Audio();
  this.sound2.src ='images/sonidos/risa.wav';
  //sprite
  this.fireSprite = new Sprite(
    "images/Tabla_Dimitri.png",
    [0, 0],
    [180, 140],
    10,
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    null,
    true,
    ctx,
    this.xpoint,
    this.ypoint,
    this.Dwidth,
    this.Dheight,
   
  );
    this.draw = function(){
      this.sprite = this.fireSprite;
      this.xpoint+=7;
      this.sprite.update(this, frames);
      this.sprite.render(ctx);
    }
}//termina fire



function Board() {
  //constructor con metodo para el lienzo
  this.x = 0;
  this.y = 0;
  this.width = canvas.width; //canvas width="1400" height="1000"
  this.height = 600; //restan 400
  this.score = 0;
  this.img = new Image();
  this.img.src = "images/scary-trees-at-night.png";
  this.img2 = new Image();
  this.img2.src = "images/rdDirt.jpg";
  this.sound = new Audio();
  this.sound.src = "images/sonidos/Tooth an Claw.MP3";
  this.sound2 = new Audio();
  this.sound2.src = "images/sonidos/begin.wav";
  this.sound3 = new Audio();
  this.sound3.src = "images/sonidos/dead.wav";

  this.img.onload = function () {
    //cuando cargue dibuja
    this.draw();
    this.drawFloor();
  }.bind(this);

  this.moveForwad = function () {
    //funcion de movimiento
    this.x--;
    if (this.x < -canvas.width) this.x = 0;
  };

  this.moveBackg = function () {
    //tengo duda ??
    this.x++;
    if (this.x < canvas.width) this.x = 0;
  };

  this.draw = function () {
    //funcion de dibujo fondo
    this.moveForwad();
    ctx.drawImage(this.img, this.x, this.y, this.width, this.height);
    ctx.drawImage(
      this.img,
      this.x + this.width,
      this.y,
      this.width,
      this.height
    );
  };

  this.drawFloor = function () {
    //funcion de dibujo piso
    this.moveForwad();
    ctx.drawImage(this.img2, this.x, 601, this.width, 400); //aarea del piso width 1400 height 400 por que es lo que resta de espacio
    ctx.drawImage(this.img2, this.x + this.width, 601, this.width, 400);
    //   ctx.drawImage(this.img2, this.x + this.width, this.y, this.width, this.height);
  };

  this.drawScore = function () {
    //funcion de score
    this.score = Math.floor(frames / 60);
    ctx.font = "50px Unquiet Spirits";
    ctx.fillStyle = "purple";
    ctx.fillText(this.score, this.width / 2, this.y + 50);
  };
} //fin del board

//enemies
function Enemies(randomPosition, randomEnemie) {
  this.health = 100;
  this.x = canvas.width; // para que salga del lado derecho del canvas
  this.y = randomPosition; //y; //de los parametros aleatorios?
  this.width = 300;
  this.height = 400; //randomHeight;
  this.img1 = new Image();
  this.img1.src = "images/enemigos/skeleton.png";
  this.img2 = new Image();
  this.img2.src = "images/enemigos/mummy.png";
  this.img3 = new Image();
  this.img3.src = "images/enemigos/zombie.png";

  this.draw = function () {
    this.x -= 5;
    switch (randomEnemie) {
      case 1:
        ctx.drawImage(this.img1, this.x, this.y, this.width, this.height);
        //ctx.strokeRect(this.x, this.y, this.width, this.height);
        break;
      case 2:
        ctx.drawImage(this.img2, this.x, this.y, this.width, this.height);
        //ctx.strokeRect(this.x, this.y, this.width, this.height);
        break;
      case 3:
        ctx.drawImage(this.img3, this.x, this.y, this.width, this.height);
        //ctx.strokeRect(this.x, this.y, this.width, this.height);
        break;
    } //final de switch
  }; //final draw

  this.isTouching = function (fires) {
    return (this.x < fires.xpoint + fires.Dwidth) && //si heroe es menor que enemigo + width
      (this.x + this.width > fires.xpoint) && //si herore + su ancho es mayor que el alto del enemigo
      (this.y< fires.ypoint + fires.Dheight) && //si el alto del heroe es menor que el alto del enemifo mas su alto
      (this.y + this.height > fires.ypoint);
  }
} //final Enemies

